src/plugin.rs

```diff
@@ -0,0 +1,218 @@
+        Python::with_gil(|py| {
+            self.0
+                .call_method1(
+                    py,
+                    pyo3::intern!(py, "confirm"),
+                    (message, yes_string, no_string),
+                )
+                .expect("`confirm` callback error")
+                .extract::<Option<bool>>(py)
+        })
+        .expect("type error in `confirm` callback")
+    }
+    fn request_public_string(&self, description: &str) -> Option<String> {
+        Python::with_gil(|py| {
+            self.0
+                .call_method1(
+                    py,
+                    pyo3::intern!(py, "request_public_string"),
+                    (description,),
+                )
+                .expect("`request_public_string` callback error")
+                .extract::<Option<String>>(py)
+        })
+        .expect("type error in `request_public_string` callback")
+    }
+    fn request_passphrase(&self, description: &str) -> Option<age::secrecy::SecretString> {
+        Python::with_gil(|py| {
+            self.0
+                .call_method1(py, pyo3::intern!(py, "request_passphrase"), (description,))
+                .expect("`request_passphrase` callback error")
+                .extract::<Option<String>>(py)
+        })
+        .expect("type error in `request_passphrase` callback")
+        .map(age::secrecy::SecretString::new)
+    }
+}
+
+#[pyclass(module = "pyrage.plugin")]
+#[derive(Clone)]
+pub(crate) struct Recipient(pub(crate) age::plugin::Recipient);
+
+#[pymethods]
+impl Recipient {
+    #[classmethod]
+    fn from_str(_cls: &PyType, v: &str) -> PyResult<Self> {
+        age::plugin::Recipient::from_str(v)
+            .map(Self)
+            .map_err(RecipientError::new_err)
+    }
+
+    fn plugin(&self) -> String {
+        self.0.plugin().to_owned()
+    }
+
+    fn __str__(&self) -> String {
+        self.0.to_string()
+    }
+}
+
+#[pyclass(module = "pyrage.plugin")]
+#[derive(Clone)]
+pub(crate) struct Identity(pub(crate) age::plugin::Identity);
+
+#[pymethods]
+impl Identity {
+    #[classmethod]
+    fn from_str(_cls: &PyType, v: &str) -> PyResult<Self> {
+        age::plugin::Identity::from_str(v)
+            .map(Self)
+            .map_err(|e| IdentityError::new_err(e.to_string()))
+    }
+
+    #[classmethod]
+    fn default_for_plugin(_cls: &PyType, plugin: &str) -> Self {
+        Self(age::plugin::Identity::default_for_plugin(plugin))
+    }
+
+    fn plugin(&self) -> String {
+        self.0.plugin().to_owned()
+    }
+
+    fn __str__(&self) -> String {
+        self.0.to_string()
+    }
+}
+
+#[pyclass(module = "pyrage.plugin")]
+#[derive(Clone)]
+pub(crate) struct RecipientPluginV1(pub(crate) Arc<age::plugin::RecipientPluginV1<PyCallbacks>>);
+
+#[pymethods]
+impl RecipientPluginV1 {
+    #[new]
+    #[pyo3(
+        text_signature = "(plugin_name: str, recipients: typing.Sequence[Recipient], identities: typing.Sequence[Identity], callbacks: Callbacks)"
```

Comment:
Does anything end up using this `text_signature`? I see it's also in `plugin.pyi`, so I'm wondering if it's a duplicate :slightly_smiling_face: